clients/drcachesim/tools/view.cpp

```diff
@@ -205,6 +157,78 @@ view_t::process_memref(const memref_t &memref)
     return parallel_shard_memref(serial_stream_, memref);
 }
 
+bool
+view_t::init_decode_cache()
+{
+    assert(decode_cache_ == nullptr);
+    decode_cache_ =
+        std::unique_ptr<decode_cache_t<disasm_info_t>>(new decode_cache_t<disasm_info_t>(
+            dcontext_.dcontext,
+            // We will decode the instr_t ourselves.
+            /*include_decoded_instr=*/false,
+            /*persist_decoded_instrs=*/false, knob_verbose_));
+    if (TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, filetype_)) {
+        error_string_ = decode_cache_->init(static_cast<offline_file_type_t>(filetype_));
+    } else if (!module_file_path_.empty()) {
+        error_string_ = decode_cache_->init(static_cast<offline_file_type_t>(filetype_),
+                                            module_file_path_, knob_alt_module_dir_);
+    } else {
+        // Continue but omit disassembly to support cases where binaries are
+        // not available and OFFLINE_FILE_TYPE_ENCODINGS is not present.
+        decode_cache_.reset(nullptr);
+    }
+    return error_string_.empty();
+}
+
+bool
+view_t::init_from_filetype()
+{
+    if (init_from_filetype_done_) {
+        return true;
+    }
+    if (filetype_ == OFFLINE_FILE_TYPE_DEFAULT) {
+        filetype_ = static_cast<offline_file_type_t>(serial_stream_->get_filetype());
+    }
+    assert(filetype_ != OFFLINE_FILE_TYPE_DEFAULT);
```

Comment:
This does not look safe: there is no documented guarantee that this won't occur, and the name implies it will occur.  Probably it does occur in some test traces: there are unit tests with trace snippets that don't have a filetype record at all.

If we want this to not be valid and be usable as a sentinel, we need to document that and change the name (by providing an alias and deprecating the old name).  And audit all the test traces and legacy traces: are you sure this isn't the type for some of those?  We would also need to add an invariant checker check, as otherwise someone could legitimately complain their trace passes our checks and so it's unreasonable to claim it's invalid here, especially without any comment at all implying this is not a well known invariant.

Reply:
This part isn't essential to this PR, so I'm going to revert it, but I'll open a new issue because it seems like a cleanup that should be done.

Reply:
https://github.com/DynamoRIO/dynamorio/issues/7228